design/dpop.md

```diff
@@ -0,0 +1,259 @@
+* ignore DPoP proofs in the “Disabled” mode.
+    
+The following options should be exposed in the adapter configuration:
+
+* DPoP Mode: required/optional/disabled
+* DPoP Proof Lifetime
+* DPoP Allowed Clock Skew
+
+> :question: Should we support DPoP for the “traditional” (monolithic) web applications, where adapters themselves handle Authorization Code flow, and the tokens are not exposed to the user agent?
+
+Classes/methods affected:
+
+*   org.keycloak.adapters.\*
+    
+
+##### Application Clustering
+
+As per the spec,
+
+> Servers SHOULD store, in the context of the request URI, the `jti` value of each DPoP proof for the time window in which the respective DPoP proof JWT would be accepted and decline HTTP requests to the same URI for which the `jti` value has been seen before.
+
+> :question: How do we implement DPoP proof replay protection in a clustered environment?  
+> 
+> Currently, only SAML adapters for Wildfly/AS/JBossWeb support distributed sessions via Infinispan.
+
+#### JavaScript Adapter
+
+In addition to DPoP handling proper, JavaScript adapter should also expose a method that would allow developers to obtain the DPoP proof header, since the latter would be required to access protected resources. As DPoP proof is bound to a particular HTTP method and URL, it should be recomputed for every protected resource request.
+
+Authors of [axios-keycloak][12], [keycloak-angular][13] and similar HTTP interceptors should be notified so that DPoP support could be introduced in the respective projects.
+
+Files affected:
+
+*   adapters/oidc/js/src/main/resources/keycloak.d.ts
+    
+
+#### Other Adapters
+
+Support in Node.JS adapter, mod\_authn\_oidc and Keycloak Gatekeeper could be also considered.
+
+### Admin UI
+
+The following configuration options should be exposed in the Admin UI for OIDC clients:
+
+* DPoP Mode: required/optional/disabled
+* DPoP Proof Lifetime
+* DPoP Allowed Clock Skew
+
+Files/classes affected:
+
+* org.keycloak.services.resources.admin.ClientResource  
+* org.keycloak.representations.idm.ClientRepresentation
+* themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html
+
+### Core
+
+Methods should be added to support [JWK thumbprint (RFC 7638)][14] computation.
+
+Classes affected:
+
+* org.keycloak.jose.jwk.{JWK,RSAPublicJWK,ECPublicJWK}
+	* add `getThumbprint()` method
+        
+
+### Tests
+
+DPoP should be properly covered by unit and integration tests.
+
+### Documentation
+
+DPoP usage should be properly documented.
+
+Affected documents:
+
+*   Securing Applications and Services Guide
+    
+## Open Questions
+
+1.  Which OAuth grant types should be DPoP-enabled?
+2.  Should we support Java adapters' redirect mode (for monolithic webapps)?
+3.  How do we implement adapter-side DPoP proof replay protection in a clustered environment?
+4.  Should we generate a DPoP keypair per session, or should it be persisted in the user agent (or both)?
```

Comment:
Basically, generating key pair per session might be good from security perspective. However, if the client can use some secure storage, per user agent might be OK.

Reply:
Agreed. We can start with a transient (per-session) keypair in JavaScript adapter, and add WebCrypto-based key storage later (will move that to milestones).